### PR TITLE
nixos/switchable-system: improve switch inhibitors

### PR DESCRIPTION
**Pull request by r-vdp**
## Things done
Improvements:
1. Turn the nix value into an attrset so that every inhibitor has a name that we can match between generations
2. Write the attrset to a file as JSON
3. When checking, we load the JSON files from both the current and the new generation into a jq pipeline and match up the keys. We output a dict with a value for every key that is present in both generations with a different value.
4. Build in error handling for different corner cases (missing files, non-JSON content)

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/tests/switch-test.nix [hunk 1]
```diff
@@ -745,6 +768,25 @@ in
...
           assert_contains(out, "this will fail")
 
+      with subtest("switch inhibitors"):
+          # Start without any inhibitors
+          switch_to_specialisation("${otherSystem}", "inhibitor0", action="switch")
+          # Check that we can switch into a generation with inhibitors from one that doesn't have any
+          switch_to_specialisation("${otherSystem}", "inhibitor1", action="switch")
+          # Check that we cannot switch into a generation that has a different value for an existing inhibitor
+          out = switch_to_specialisation("${otherSystem}", "inhibitor2", action="switch", fail=True)
+          print(out)
```
Comment (jmbaur) [inline nixos/tests/switch-test.nix]:
```suggestion
```

File: nixos/tests/switch-test.nix [hunk 2]
```diff
@@ -656,12 +656,35 @@ in
...
+        inhibitor2.configuration.system.switch.inhibitors = lib.mkForce {
+          foo = "baz";
+          quz = "boz";
+        };
+
+        inhibitor3.configuration.system.switch.inhibitors = lib.mkForce {
+          foo = "bar";
+          qux = "baz";
+        };
+      };
```
Comment (jmbaur) [inline nixos/tests/switch-test.nix]: Can you move this into a single specialisation under the `machine` node? It's not clear to me why this `other` node exists, it seems like a misc dumping ground. As for moving the inhibitor testing under single specialisation, this test is _extremely_ expensive to evaluate given the large number of specialisations, I would personally like to keep that number lower, even if just to make iterating on it less painful. Also, if possible, can you choose different names for the test's inhibitors themselves? It's not  clear to me reading the assertions in `testScript` what this is supposed to be doing just by the "foo"/"bar"/"baz"/etc.

Comment (r-vdp) [reply nixos/tests/switch-test.nix]: OK, we need the different specialisations though I think, because otherwise how would we test switching between them?

Comment (r-vdp) [reply nixos/tests/switch-test.nix]: I moved the specialisations to the `machine` node. I think we need multiple ones though since we want to test the behaviour when we switch between them. If you have another way of doing this in mind that can reduce the amount of specialisations, please let me know.

Comment (r-vdp) [reply nixos/tests/switch-test.nix]: Also, the test takes exactly one minute to evaluate on my zen4 laptop, so I don't think it's that bad.
Eval times improved a lot when we switched the test over to use `runTest`.

Comment (r-vdp) [reply nixos/tests/switch-test.nix]: Also, the test takes exactly one minute to evaluate on my zen4 laptop, so I don't think it's that bad.
Eval times improved a lot when we switched the test over to use `runTest`.

Comment (jmbaur) [reply nixos/tests/switch-test.nix]: I was more so talking about how much memory it takes to evaluate every single nixos machine variant with all the specialisations. Each one that is added makes it more painful, but I see now why this is needed for the use case here, you're testing the change of critical packages _between_ switches. The foo/bar/baz naming still doesn't make that super clear, but I guess the keys and values can really be anything, so not much to do there.

File: nixos/modules/services/system/dbus.nix [hunk 3]
```diff
@@ -104,6 +104,8 @@ in
 
   config = mkIf cfg.enable (mkMerge [
     {
+      system.switch.inhibitors.dbus-implementation = cfg.implementation;
```
[the remark this answers is not in the document]
Comment (r-vdp) [reply nixos/modules/services/system/dbus.nix]: Sure, I'll do that tonight.

**Review by jmbaur — APPROVED** [review]

**Review by arianvp — APPROVED** [review]
Great stuff. We probably also want the package version to be an inhibitor though?
